packages/core-p2p/src/peer-verifier.ts

```diff
@@ -0,0 +1,553 @@
+            // ourBlocksHeights[2] = 1200, the peer does not have this block
+            // ourBlocksHeights[3] = 1300, the peer does not have this block
+            // we will get indexOfHighestCommon = 1, and for the next iteration:
+            // low = 1100 + 1 = 1101
+            // high = ourBlocksHeights[1 + 1] - 1 = ourBlocksHeights[2] - 1 = 1200 - 1 = 1199
+            const indexOfHighestCommon = ourBlocksHeights.indexOf(highestCommonBlockHeight);
+            assert.notStrictEqual(indexOfHighestCommon, -1);
+            assert(indexOfHighestCommon < ourBlocksHeights.length - 1);
+            low = highestCommonBlockHeight + 1;
+            high = ourBlocksHeights[indexOfHighestCommon + 1] - 1;
+        }
+
+        if (highestCommonBlockHeight === null) {
+            this.logger.info(`${this.logPrefix} failure: could not determine highest common block`);
+        } else {
+            this.logger.debug(`${this.logPrefix} definitive highest common: ${highestCommonBlockHeight}`);
+        }
+
+        return highestCommonBlockHeight;
+    }
+
+    /**
+     * Verify the blocks of the peer's chain that are in the range [height, last block in round].
+     * @param {Number} height verify blocks at and after this height
+     * @param {Boolean} includeHisBlocksInOurChain indicates whether we should advance our
+     * chain with the peer's blocks should they pass validation
+     * @return {Boolean} true if the blocks are legit (signed by the appropriate delegates)
+     */
+    private async verifyPeerBlocks(height: number, includeHisBlocksInOurChain: boolean): Promise<boolean> {
+        const round = roundCalculator.calculateRound(height);
+        const lastBlockHeightInRound = round.round * round.maxDelegates;
+
+        // Verify a few blocks that are not too far up from the last common block. Within the
+        // same round as the last common block or in the next round if the last common block is
+        // the last block in a round (so that the delegates calculations are still the same for
+        // both chains).
+
+        const delegates = await this.getDelegates(round);
+
+        const hisBlocksByHeight = {};
+
+        for (let h = height; h <= lastBlockHeightInRound; h++) {
+            if (hisBlocksByHeight[h] === undefined) {
+                if (!await this.fetchBlocksFromHeight(h, hisBlocksByHeight)) {
+                    return false;
+                }
+            }
+            assert(hisBlocksByHeight[h] !== undefined);
+
+            const hisBlockJson = hisBlocksByHeight[h];
+
+            const hisBlockObj = await this.verifyPeerBlock(hisBlockJson, h, delegates);
+            if (hisBlockObj === null) {
+                return false;
+            }
+
+            if (includeHisBlocksInOurChain) {
+                // XXX this code is the same as
+                // packages/core-p2p/src/server/versions/1/handlers.ts:postBlock
+
+                /* XXX this.blockchain is null here
```

Comment:
Figure out why blockchain is null. Maybe during startup this code is executed before the blockchain is started. If so, see if it is possible to start the blockchain earlier or if not, exec this code only if not null (e.g. only during pings after startup has completed).

Reply:
Yes, blockchain is started after p2p and it would be too involved to swap the order as it would require removing the p2p dependencies from blockchain.